src/library_dylink.js

```diff
@@ -270,51 +270,114 @@ var LibraryDylink = {
-
-    // shared libraries this module needs. We need to load them first, so that
-    // current module could resolve its imports. (see tools/shared.py
-    // WebAssembly.make_shared_library() for "dylink" section extension format)
-    var neededDynlibsCount = getLEB();
-    var neededDynlibs = [];
-    for (var i = 0; i < neededDynlibsCount; ++i) {
-      var nameLen = getLEB();
-      var nameUTF8 = binary.subarray(next, next + nameLen);
-      next += nameLen;
-      var name = UTF8ArrayToString(nameUTF8, 0);
-      neededDynlibs.push(name);
+
+    function parseDylinkSection(byteArray) {
+      var customSection = {
+        memorySize : -1,
+        memoryAlign : -1,
+        tableSize : -1,
+        tableAlign : -1,
+        neededDynlibs: []
+      }
+
+      customSection.memorySize = getLEB(byteArray);
+      customSection.memoryAlign = getLEB(byteArray);
+      customSection.tableSize = getLEB(byteArray);
+      customSection.tableAlign = getLEB(byteArray);
+      // shared libraries this module needs. We need to load them first, so that
+      // current module could resolve its imports. (see tools/shared.py
+      // WebAssembly.make_shared_library() for "dylink" section extension format)
+      var neededDynlibsCount = getLEB(byteArray);
+      for (var i = 0; i < neededDynlibsCount; ++i) {
+        var nameLen = getLEB(byteArray);
+        var nameUTF8 = byteArray.subarray(next, next + nameLen);
+        next += nameLen;
+        var name = UTF8ArrayToString(nameUTF8, 0);
+        customSection.neededDynlibs.push(name);
+      }
+      return customSection;
+    }
+
+    if (binary instanceof WebAssembly.Module) {
+      var dylinkSection = WebAssembly.Module.customSections(binary, "dylink");
+      assert(dylinkSection.length != 0, 'need dylink section');
+      var dylinkSectionInt8 = new Int8Array(dylinkSection[0]);
+      return parseDylinkSection(dylinkSectionInt8);
+    } else {
+      var int32View = new Uint32Array(new Uint8Array(binary.subarray(0, 24)).buffer);
+      assert(int32View[0] == 0x6d736100, 'need to see wasm magic number'); // \0asm
+      // we should see the dylink section right after the magic number and wasm version
+      assert(binary[8] === 0, 'need the dylink section to be first')
+      next = 9;
+      getLEB(binary); //section size
+      assert(binary[next] === 6);                 next++; // size of "dylink" string
+      assert(binary[next] === 'd'.charCodeAt(0)); next++;
+      assert(binary[next] === 'y'.charCodeAt(0)); next++;
+      assert(binary[next] === 'l'.charCodeAt(0)); next++;
+      assert(binary[next] === 'i'.charCodeAt(0)); next++;
+      assert(binary[next] === 'n'.charCodeAt(0)); next++;
+      assert(binary[next] === 'k'.charCodeAt(0)); next++;
+
+      return parseDylinkSection(binary);
+    }
+  },
+
+  // Module.symbols <- libModule.symbols (flags.global handler)
+  $mergeLibSymbols__deps: ['$asmjsMangle'],
+  $mergeLibSymbols: function(libModule, lib) {
```

Comment:
Perhaps the second arguments could be better named?  libName?

Reply:
Actually better still maybe just pass the `dso` object which hold both of these?

Reply:
the caller from outside the library might not know what to put into the object, i think its clearer this way with the 2 args.